### PR TITLE
[MM-23450] Deploy a clustered load test agent setup using Terraform

### PR DESCRIPTION
**Pull request by isacikgoz**
#### Summary
This PR adds loadtest agents and a coordinator while running load tests on AWS.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-23450

### CODE REVIEW

File: deployment/config.go [hunk 1]
```diff
@@ -19,6 +19,9 @@ import (
 type Config struct {
 	ClusterName      string // Name of the cluster.
 	AppInstanceCount int    // Number of application instances.
+	// Number of agents; at least 2 agents required so that one of them will work
+	// as a coordinator.
+	AgentCount       int
```
Comment (agnivade) [inline deployment/config.go]: I was thinking to use the coordinator in one of the machines as agent. A coordinator does not consume any extra resources, so it's fine to use it on the same machine.

File: deployment/terraform/cluster.tf [hunk 2]
```diff
@@ -164,6 +197,32 @@ resource "aws_security_group" "db" {
...
+    to_port     = 22
+    protocol    = "tcp"
+    cidr_blocks = ["0.0.0.0/0"]
+  }
+
+  ingress {
+    from_port   = 4000
+    to_port     = 4000
+    protocol    = "tcp"
+    cidr_blocks = ["0.0.0.0/0"]
```
Comment (agnivade) [inline deployment/terraform/cluster.tf]: Make it accessible only within the agent group.

File: deployment/terraform/cluster.tf [hunk 3]
```diff
@@ -108,6 +108,39 @@ resource "aws_rds_cluster" "db_cluster" {
...
+      inline = [
+      "while [ ! -f /var/lib/cloud/instance/boot-finished ]; do echo 'Waiting for cloud-init...'; sleep 1; done",
+      "wget https://dl.google.com/go/${var.go_binary_file}",
+      "sudo tar -C /usr/local -xzf ${var.go_binary_file}",
+      "wget https://github.com/mattermost/mattermost-load-test-ng/archive/${var.loadtest_source_code_ref}.tar.gz",
+      "tar -xzf ${var.loadtest_source_code_ref}.tar.gz",
+      "cp mattermost-load-test-ng-${var.loadtest_source_code_ref}/config/simplecontroller.default.json simplecontroller.json",
+      "cp mattermost-load-test-ng-${var.loadtest_source_code_ref}/config/config.default.json config.json",
+      "cp -r mattermost-load-test-ng-${var.loadtest_source_code_ref}/testdata testdata",
+      "cd mattermost-load-test-ng-${var.loadtest_source_code_ref} && export PATH=$PATH:/usr/local/go/bin && go build -o agent ./cmd/loadtest && sudo mv agent /usr/local/bin"
```
Comment (agnivade) [inline deployment/terraform/cluster.tf]: Building binaries won't work if I am not wrong. That's what we have been trying to avoid. But we have all the hardcodings to profile images and testdata directories. So moving the binary to some place else will fail. Let's just do a "go run ./cmd/loadtest" and "go run ./cmd/coordinator".

File: deployment/terraform/create.go [hunk 4]
```diff
@@ -168,6 +183,41 @@ func (t *Terraform) Create() error {
...
+	// TODO: replace with reverse nginx ip
+	sshc, err := extAgent.NewClient(output.Instances.Value[0].PublicIP)
+	if err != nil {
+		return err
+	}
+	if err := sshc.RunCommand(cmd); err != nil {
+		return err
+	}
+
+	for i := 1; i < len(output.Agents.Value); i++ {
```
Comment (agnivade) [inline deployment/terraform/create.go]: We can go from 0 here.

File: deployment/config.go [hunk 5]
```diff
@@ -77,6 +88,9 @@ func ReadConfig(filePath string) (*Config, error) {
 	v.SetDefault("LogSettings.FileJson", true)
 	v.SetDefault("LogSettings.FileLocation", "loadtest.log")
 
+	v.SetDefault("GoBinaryFile", "go1.14.1.linux-amd64.tar.gz")
```
Comment (agnivade) [inline deployment/config.go]: Let's just take "1.14.1". Rest can be concatenated.

File: deployment/config.go [hunk 6]
```diff
@@ -32,6 +35,11 @@ type Config struct {
 	// stable release.
 	MattermostDownloadURL string
 	MattermostLicenseFile string // Path to the Mattermost EE license file.
+	AdminEmail            string // Mattermost instance sysadmin e-mail.
+	AdminUsername         string // Mattermost instance sysadmin user name.
+	AdminPassword         string // Mattermost instance sysadmin password.
+	GoBinaryFile          string // Go binaries to compile loadtest-agents.
+	SourceCodeRef         string // loadtest-ng head reference
```
Comment (agnivade) [inline deployment/config.go]: How about naming it branch ?

Comment (isacikgoz) [reply deployment/config.go]: We can also target tags.

File: deployment/config.go (hunk 1 above)
Comment (streamer45) [reply deployment/config.go]: From a theoretical point of view I would prefer the coordinator to be separated from the agents it controls.  

A server running an agent with many controllers/users might indeed experience a lot of pressure. The coordinator completes the feedback loop and it's important that it can act as fast as possible when data comes in from prometheus.

That said, it's probably fine for now to have a shared instance (load-agent and coordinator) and see how that works.

The coordinator could also be run locally by the user doing the load-test since it's the initiator.

Comment (agnivade) [reply deployment/config.go]: Hmm .. fine by me.

File: deployment/terraform/create.go [hunk 7]
```diff
@@ -181,8 +220,34 @@ func (t *Terraform) setupAppServers(output *terraformOutput, extAgent *ssh.ExtAg
 	}
 }
 
+func (t *Terraform) setupLoadtestAgents(extAgent *ssh.ExtAgent, output *terraformOutput) error {
+	// Creating the sysadmin user.
+	// should wait the server to start
+
+	for i := 0; i < len(output.Agents.Value); i++ {
```
Comment (agnivade) [inline deployment/terraform/create.go]: Nit: any reason we are not using a range here ?

File: deployment/terraform/create.go [hunk 8]
```diff
@@ -181,8 +220,34 @@ func (t *Terraform) setupAppServers(output *terraformOutput, extAgent *ssh.ExtAg
...
+		val := output.Agents.Value[i]
+
+		if err := t.runAgent(extAgent, val.PublicIP); err != nil {
+			mlog.Error("error while setting up an agent", mlog.String("agent name", val.Tags.Name), mlog.Err(err))
+			continue
+		}
+	}
+
+	if err := t.updateCoordinatorConfig(extAgent, output); err != nil {
+		mlog.Error(err.Error())
```
Comment (agnivade) [inline deployment/terraform/create.go]: Let's return an error instead of logging here ? You are not following your own suggestions :)

File: deployment/terraform/create.go [hunk 9]
```diff
@@ -181,8 +220,34 @@ func (t *Terraform) setupAppServers(output *terraformOutput, extAgent *ssh.ExtAg
...
 
+func (t *Terraform) setupLoadtestAgents(extAgent *ssh.ExtAgent, output *terraformOutput) error {
+	// Creating the sysadmin user.
+	// should wait the server to start
+
+	for i := 0; i < len(output.Agents.Value); i++ {
+		val := output.Agents.Value[i]
+
+		if err := t.runAgent(extAgent, val.PublicIP); err != nil {
+			mlog.Error("error while setting up an agent", mlog.String("agent name", val.Tags.Name), mlog.Err(err))
```
Comment (agnivade) [inline deployment/terraform/create.go]: Same as below, let's return the error instead of continuing. We want the entire thing to be an atomic and idempotent operation. So if it fails, the user can try it again. But when it passes, that means everything has passed perfectly.

File: deployment/terraform/agent.go [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+	mlog.Info("Populating DB")
+	cmd := fmt.Sprintf("cd mattermost-load-test-ng-%s && export PATH=$PATH:/usr/local/go/bin && go run ./cmd/loadtest init", t.config.SourceCodeRef)
+	if err := sshc.RunCommand(cmd); err != nil {
+		return err
+	}
+
+	// Starting coordinator.
+	mlog.Info("Starting coordinator", mlog.String("ip", ip))
+	cmd = fmt.Sprintf("cd mattermost-load-test-ng-%s && export PATH=$PATH:/usr/local/go/bin && go run ./cmd/coordinator", t.config.SourceCodeRef)
+	go func() {
```
Comment (agnivade) [inline deployment/terraform/agent.go]: Mmm .. I liked the `&` approach better. The problem with this is that if the command fails in any way, we wouldn't know. We want to be absolutely sure that everything has passed when we say "deployment complete".

Later, all of this should move to their own systemd service, at which point the `&` hack would be unncessary.

File: deployment/terraform/agent.go [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+	// Populate the DB.
+	mlog.Info("Populating DB")
+	cmd := fmt.Sprintf("cd mattermost-load-test-ng-%s && export PATH=$PATH:/usr/local/go/bin && go run ./cmd/loadtest init", t.config.SourceCodeRef)
+	if err := sshc.RunCommand(cmd); err != nil {
+		return err
+	}
+
+	// Starting coordinator.
+	mlog.Info("Starting coordinator", mlog.String("ip", ip))
+	cmd = fmt.Sprintf("cd mattermost-load-test-ng-%s && export PATH=$PATH:/usr/local/go/bin && go run ./cmd/coordinator", t.config.SourceCodeRef)
```
Comment (agnivade) [inline deployment/terraform/agent.go]: Assuming that the master branch thing is taken care of, we can write it like this:
`cd mattermost-load-test-ng && PATH=$PATH:/usr/local/go/bin go run ./cmd/coordinator &`.

File: deployment/terraform/agent.go [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+
+func (t *Terraform) runAgent(extAgent *ssh.ExtAgent, ip string) error {
+	sshc, err := extAgent.NewClient(ip)
+	if err != nil {
+		return err
+	}
+
+	// Starting agent.
+	mlog.Info("Starting agent", mlog.String("ip", ip))
+	cmd := fmt.Sprintf("cd mattermost-load-test-ng-%s && export PATH=$PATH:/usr/local/go/bin && go run ./cmd/loadtest server", t.config.SourceCodeRef)
```
Comment (agnivade) [inline deployment/terraform/agent.go]: Same 2 comments as above.

File: deployment/terraform/agent.go [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+}
+
+func (t *Terraform) startCoordinator(extAgent *ssh.ExtAgent, ip string) error {
+	sshc, err := extAgent.NewClient(ip)
+	if err != nil {
+		return err
+	}
+
+	// Populate the DB.
+	mlog.Info("Populating DB")
```
Comment (agnivade) [inline deployment/terraform/agent.go]: Let's also add the `ip` tag.

File: deployment/terraform/agent.go [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+				},
+			},
+		},
+	}
+
+	data, err := json.Marshal(loadtestConfig)
+	if err != nil {
+		return err
+	}
+	dstPath := fmt.Sprintf("/home/ubuntu/mattermost-load-test-ng-%s/config/config.json", t.config.SourceCodeRef)
```
Comment (agnivade) [inline deployment/terraform/agent.go]: Same as earlier, let's remove "-master".

File: deployment/terraform/cluster.tf [hunk 15]
```diff
@@ -143,6 +143,38 @@ resource "aws_rds_cluster" "db_cluster" {
...
+  provisioner "remote-exec" {
+      inline = [
+      "while [ ! -f /var/lib/cloud/instance/boot-finished ]; do echo 'Waiting for cloud-init...'; sleep 1; done",
+      "wget https://dl.google.com/go/go${var.go_version}.linux-amd64.tar.gz",
+      "sudo tar -C /usr/local -xzf go${var.go_version}.linux-amd64.tar.gz",
+      "sudo sh -c \"echo 'export PATH=$PATH:/usr/local/go/bin' >> /etc/profile\"",
+      "wget https://github.com/mattermost/mattermost-load-test-ng/archive/${var.loadtest_source_code_ref}.tar.gz",
+      "tar -xzf ${var.loadtest_source_code_ref}.tar.gz",
+      "cp mattermost-load-test-ng-${var.loadtest_source_code_ref}/config/simplecontroller.default.json mattermost-load-test-ng-${var.loadtest_source_code_ref}/config/simplecontroller.json",
+      "cp mattermost-load-test-ng-${var.loadtest_source_code_ref}/config/config.default.json mattermost-load-test-ng-${var.loadtest_source_code_ref}/config/config.json",
```
Comment (agnivade) [inline deployment/terraform/cluster.tf]: We are already overwriting this later. Do we still need to do this ?

File: deployment/terraform/create.go [hunk 16]
```diff
@@ -127,6 +147,25 @@ func (t *Terraform) Create() error {
 	// Updating the nginx config on proxy server
 	t.setupProxyServer(output, extAgent)
 
+	time.Sleep(30 * time.Second)
```
Comment (agnivade) [inline deployment/terraform/create.go]: Why is this needed ? Did you face any issues where mattermost wasn't yet up when you were trying to create the sysadmin ?

File: deployment/terraform/create.go [hunk 17]
```diff
@@ -181,8 +220,34 @@ func (t *Terraform) setupAppServers(output *terraformOutput, extAgent *ssh.ExtAg
...
+			continue
+		}
+	}
+
+	if err := t.updateCoordinatorConfig(extAgent, output); err != nil {
+		mlog.Error(err.Error())
+		// return err
+	}
+
+	if err := t.startCoordinator(extAgent, output.Agents.Value[0].PublicIP); err != nil {
```
Comment (agnivade) [inline deployment/terraform/create.go]: I am still thinking whether we want to start the coordinator immediately with the deployment. This means a deployment will essentially start running the load test.

I was thinking to have a separate command for it like "./cmd/deployer run", which would get the *terraformOutput again and just call this method.

Logically, running the load test is a separate step from deployment IMO. If the operator wants to run them both, then they can just `&&` the commands.

Not strongly opinionated on this though. Any thoughts ?

File: deployment/config.go (hunk 1 above)
Comment (agnivade) [reply deployment/config.go]: We are going with same machine for now. Slating this for a later improvement.

File: deployment/terraform/agent.go [hunk 18]
```diff
@@ -0,0 +1,140 @@
...
+	"github.com/mattermost/mattermost-load-test-ng/coordinator/performance/prometheus"
+	"github.com/mattermost/mattermost-load-test-ng/deployment/terraform/ssh"
+	"github.com/mattermost/mattermost-load-test-ng/loadtest"
+	"github.com/mattermost/mattermost-server/v5/mlog"
+)
+
+func (t *Terraform) generateLoadtestAgentConfig(output *terraformOutput) loadtest.Config {
+	return loadtest.Config{
+		ConnectionConfiguration: loadtest.ConnectionConfiguration{
+			// TODO: replace with reverse nginx ip
```
[the remark this answers is not in the document]
Comment (isacikgoz) [reply deployment/terraform/agent.go]: Ah, it escaped me :)

File: deployment/terraform/cluster.tf [hunk 19]
```diff
@@ -143,6 +143,38 @@ resource "aws_rds_cluster" "db_cluster" {
...
+  vpc_security_group_ids = ["${aws_security_group.agent.id}"]
+
+  provisioner "remote-exec" {
+      inline = [
+      "while [ ! -f /var/lib/cloud/instance/boot-finished ]; do echo 'Waiting for cloud-init...'; sleep 1; done",
+      "wget https://dl.google.com/go/go${var.go_version}.linux-amd64.tar.gz",
+      "sudo tar -C /usr/local -xzf go${var.go_version}.linux-amd64.tar.gz",
+      "sudo sh -c \"echo 'export PATH=$PATH:/usr/local/go/bin' >> /etc/profile\"",
+      "wget https://github.com/mattermost/mattermost-load-test-ng/archive/${var.loadtest_source_code_ref}.tar.gz",
+      "tar -xzf ${var.loadtest_source_code_ref}.tar.gz",
```
[the remark this answers is not in the document]
Comment (isacikgoz) [reply deployment/terraform/cluster.tf]: Definitely! 😅

File: deployment/terraform/create.go [hunk 20]
```diff
@@ -127,6 +147,25 @@ func (t *Terraform) Create() error {
 	// Updating the nginx config on proxy server
 	t.setupProxyServer(output, extAgent)
 
+	time.Sleep(30 * time.Second)
+	cmd := fmt.Sprintf("/opt/mattermost/bin/mattermost user create --email %s --username %s --password %s --system_admin",
```
[the remark this answers is not in the document]
Comment (isacikgoz) [reply deployment/terraform/create.go]: Done 👍

File: deployment/terraform/create.go (hunk 7 above)
Comment (isacikgoz) [reply deployment/terraform/create.go]: Hmm, no reason. It was starting from 1 in the earlier version of this PR. Changed to range 👍

File: deployment/terraform/create.go (hunk 8 above)
Comment (isacikgoz) [reply deployment/terraform/create.go]: Yeah, it's like a "the shoemaker's son always goes barefoot" 😄

File: deployment/terraform/create.go (hunk 9 above)
Comment (isacikgoz) [reply deployment/terraform/create.go]: Fixed 👍

File: deployment/terraform/agent.go (hunk 10 above)
Comment (isacikgoz) [reply deployment/terraform/agent.go]: I tried `&` but it hangs on because; `crypto/ssh` runs the command as `bash -c cd mattermost-load-test-ng-master && go run ./cmd/loadtest server &` and to be able to make `&` work, we need to put entire command into quotes(") but in that case it failed again, didn't want to spend too much time on it TBH. We could simply use something like `session.Start(..)`. Maybe we can address that in a seperate PR. I can take the initiative 👍

File: deployment/terraform/agent.go (hunk 12 above)
Comment (isacikgoz) [reply deployment/terraform/agent.go]: Done 👍

File: deployment/terraform/agent.go (hunk 13 above)
Comment (isacikgoz) [reply deployment/terraform/agent.go]: Done. 👍

File: deployment/terraform/agent.go (hunk 14 above)
Comment (isacikgoz) [reply deployment/terraform/agent.go]: Done. 👍

File: deployment/terraform/cluster.tf (hunk 15 above)
Comment (isacikgoz) [reply deployment/terraform/cluster.tf]: We are overwriting for the coordinator only, but we start normal agents before we configure coordinator. And the need for this will be vanished after I send the fix 👍

File: deployment/terraform/create.go (hunk 16 above)
Comment (isacikgoz) [reply deployment/terraform/create.go]: Yes, I was getting `109` error return or smth like that. After a brief diving into mm store code, it means `cannot create index error`. But I don't remember if I tested after this patch 🤔 Will test it 👍

File: deployment/terraform/create.go (hunk 17 above)
Comment (isacikgoz) [reply deployment/terraform/create.go]: In my opinion, since we are setting everything for loadtesting, it *should* start straight. But if we need to do something before running the coordinator, it would make sense. What could be the reason not to start load test?

File: deployment/terraform/agent.go (hunk 10 above)
Comment (agnivade) [reply deployment/terraform/agent.go]: I see. Thanks for debugging this. Yeah let's take this on a separate PR.

File: deployment/terraform/create.go [hunk 21]
```diff
@@ -295,6 +277,25 @@ func (t *Terraform) setupProxyServer(output *terraformOutput, extAgent *ssh.ExtA
 	}()
 }
 
+func (t *Terraform) createAdminUser(extAgent *ssh.ExtAgent, output *terraformOutput) error {
+	time.Sleep(30 * time.Second)
```
Comment (agnivade) [inline deployment/terraform/create.go]: Let's move the sleep to the main method.

File: deployment/terraform/create.go (hunk 17 above)
Comment (agnivade) [reply deployment/terraform/create.go]: > What could be the reason not to start load test?

Actually, thinking more about it. We would probably reuse the same environment to run multiple load tests with different versions of mattermost. In that case, after running a test, we would upload a different binary and run the test again. So we might need a separate command to stop/start the load test. 

Hence, I was thinking that deployment just deploys. And then we have a separate command that starts/stop the coordinator.

File: deployment/terraform/create.go (hunk 9 above)
Comment (agnivade) [reply deployment/terraform/create.go]: I still see the continue statement :thinking:

File: deployment/terraform/create.go (hunk 17 above)
Comment (isacikgoz) [reply deployment/terraform/create.go]: Sounds reasonable. 👍

File: deployment/terraform/create.go (hunk 16 above)
Comment (agnivade) [reply deployment/terraform/create.go]: That's okay for now. But let's have a check that pings the `curl http://localhost:8065/api/v4/system/ping` endpoint and checks for a proper response before proceeding. Created https://mattermost.atlassian.net/browse/MM-23566 for this.

**Comment by agnivade** [comment]
Oh I also forgot, please stick in a `"sudo apt-get -y update",` in the remote-exec for the load_test_agent. :)

**Comment by agnivade** [comment]
Found a bug during testing: `updateCoordinatorConfig` only updates the agent config and coordinator config of the first agent. We should update agent config on all agents, and coordinator config on the first.

That's why agent wasn't getting started on all machines.

Please also add logging to the agents, so that any errors get logged for debugging.

```diff
iff --git a/deployment/terraform/agent.go b/deployment/terraform/agent.go
index e7cd7ce..396782f 100644
--- a/deployment/terraform/agent.go
+++ b/deployment/terraform/agent.go
@@ -12,6 +12,7 @@ import (
        "github.com/mattermost/mattermost-load-test-ng/coordinator/performance/prometheus"
        "github.com/mattermost/mattermost-load-test-ng/deployment/terraform/ssh"
        "github.com/mattermost/mattermost-load-test-ng/loadtest"
+       "github.com/mattermost/mattermost-load-test-ng/logger"
        "github.com/mattermost/mattermost-server/v5/mlog"
 )
 
@@ -35,6 +36,11 @@ func (t *Terraform) generateLoadtestAgentConfig(output *terraformOutput) loadtes
                        InitialActiveUsers: 4,
                        MaxActiveUsers:     1000,
                },
+               LogSettings: logger.Settings{
+                       EnableFile:   true,
+                       FileLevel:    "INFO",
+                       FileLocation: "loadtest.log",
+               },
        }
 }

```

**Comment by agnivade** [comment]
Going to test this soon. I think it looks great now. :)

**Comment by agnivade** [comment]
Ok tested it, and everything looks good. Only thing that I noticed is that the coordinator didn't start. And I also didn't see any logs related to it. Other than that, all agents were started and running. No other issues noticed. Init also happened fine.

In your testing, did the coordinator start fine ?

File: deployment/terraform/create.go [hunk 22]
```diff
@@ -223,6 +223,7 @@ func (t *Terraform) setupLoadtestAgents(extAgent *ssh.ExtAgent, output *terrafor
 		return err
 	}
 
+	time.Sleep(30 * time.Second)
```
Comment (isacikgoz) [inline deployment/terraform/create.go]: Well, this is a precaution, not that elegant but good to have.

File: deployment/terraform/agent.go [hunk 23]
```diff
@@ -0,0 +1,143 @@
...
+		return err
+	}
+	dstPath := "/home/ubuntu/mattermost-load-test-ng/config/coordinator.json"
+	if err := sshc.Upload(bytes.NewReader(data), dstPath, false); err != nil {
+		return fmt.Errorf("error running ssh command: %w", err)
+	}
+
+	// Starting coordinator.
+	mlog.Info("Compiling coordinator", mlog.String("ip", ip))
+	cmd := "cd mattermost-load-test-ng && export PATH=$PATH:/usr/local/go/bin && go build -o lt-coordinator ./cmd/coordinator"
```
Comment (agnivade) [inline deployment/terraform/agent.go]: Just add a TODO comment here saying that this is a hack for now.

File: deployment/terraform/agent.go [hunk 24]
```diff
@@ -0,0 +1,143 @@
...
+	data, err := json.Marshal(clusterConfig)
+	if err != nil {
+		return err
+	}
+	dstPath := "/home/ubuntu/mattermost-load-test-ng/config/coordinator.json"
+	if err := sshc.Upload(bytes.NewReader(data), dstPath, false); err != nil {
+		return fmt.Errorf("error running ssh command: %w", err)
+	}
+
+	// Starting coordinator.
```
Comment (isacikgoz) [inline deployment/terraform/agent.go]:
```suggestion
	// TODO: This is a hack to overcome an issue with go run command.
```

File: deployment/config.go [hunk 25]
```diff
@@ -32,6 +33,11 @@ type Config struct {
 	// stable release.
 	MattermostDownloadURL string
 	MattermostLicenseFile string // Path to the Mattermost EE license file.
+	AdminEmail            string // Mattermost instance sysadmin e-mail.
+	AdminUsername         string // Mattermost instance sysadmin user name.
+	AdminPassword         string // Mattermost instance sysadmin password.
+	GoVersion             string // Go version to download for compiling loadtest-agents.
+	SourceCodeRef         string // loadtest-ng head reference
```
Comment (streamer45) [inline deployment/config.go]: Missing dot at the end.

File: deployment/config.go [hunk 26]
```diff
@@ -53,6 +59,9 @@ func (c *Config) IsValid() error {
 	if len(clusterName) == 0 || !unicode.IsLetter(firstRune) || !isAlphanumeric(clusterName) {
 		return fmt.Errorf("db cluster name must begin with a letter and contain only alphanumeric characters")
 	}
+	if c.AgentCount < 2 {
+		return fmt.Errorf("number of agents must be greater than 2")
+	}
```
Comment (streamer45) [inline deployment/config.go]: Is there a technical reason for this? Everything can work fine with 1 agent as far as I know.

Comment (isacikgoz) [reply deployment/config.go]: This one remained from the initial approach we can change this to `1` 👍

File: deployment/config.go [hunk 27]
```diff
@@ -19,6 +19,7 @@ import (
 type Config struct {
 	ClusterName      string // Name of the cluster.
 	AppInstanceCount int    // Number of application instances.
+	AgentCount       int    // Number of agents, first agent will also be the coordinator.
```
Comment (streamer45) [inline deployment/config.go]:
```suggestion
	AgentCount       int    // Number of agents, first agent and coordinator will share the same instance.
```
